### PR TITLE
Network Simulator: class Flow

### PR DESCRIPTION
**Pull request by AntoxaBarin**
Closes Issue https://github.com/Cloud-Storage-Team/algnet/issues/9. Implements class Flow: a top-level user interface which describes a socket pair (source and destination).

### CODE REVIEW

**Review by drgulevich — COMMENTED** [review]
> Files to review: [include/Flow.hpp](https://github.com/Cloud-Storage-Team/algnet/pull/17/files#diff-8681edf1215654fc2070f67860e21579e7d7e95159880801bea914393b830de7), [src/Flow.cpp](https://github.com/Cloud-Storage-Team/algnet/pull/17/files#diff-4a9770ab104dff163033f079dc1863dc13faa9630353c9efe071510cf5a4e39e).

Can we have only relevant files which close issue 9 here, please?

**Comment by drgulevich** [comment]
> Files to review: [include/Flow.hpp](https://github.com/Cloud-Storage-Team/algnet/pull/17/files#diff-8681edf1215654fc2070f67860e21579e7d7e95159880801bea914393b830de7), [src/Flow.cpp](https://github.com/Cloud-Storage-Team/algnet/pull/17/files#diff-4a9770ab104dff163033f079dc1863dc13faa9630353c9efe071510cf5a4e39e). Issue #9.

The PR should say "Closes Issue https://github.com/Cloud-Storage-Team/algnet/issues/9." This way it will close the issue automatically after the merge is complete.

File: include/Server.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     * Get server's current time in nanoseconds
+     * @return std::uint64_t -- server's current time in nanoseconds
+     */
+    std::uint64_t GetCurrentTime() const;
+
+    /**
+     * Replace server's current time with a new value
+     * @param new_time -- new value of time in nanoseconds
+     */
+    void SetCurrentTime(std::uint64_t new_time);
```
Comment (drgulevich) [inline include/Server.hpp:35]: May I, please, ask not to use getters/setters at this stage. See the opinion on this here https://youtu.be/uHSLHvWFkto?t=4380 discussion and further discussion https://stackoverflow.com/questions/32625274/public-variables-bad-practice-vs-getters-and-setters-functions.

File: include/Flow.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * @brief Server receiver object
+     */
+    ServerBase destination_node;
+
+    /**
+     * @brief Priority queue with in-flight packets
+     */
+    std::priority_queue<Packet> packets;
```
Comment (drgulevich) [inline include/Flow.hpp:42]: Why does this priority queue belong to the flow? What is the meaning of this priority queue?

File: include/Flow.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+    * @brief Last ID given to network device.
+    */
+    std::uint32_t last_given_id = 0;
+
+    /**
+     * @brief Vector with sending server objects
+     */
+    std::vector<ServerSender> source_nodes;
```
Comment (drgulevich) [inline include/Flow.hpp:32]: Why does a flow correspond to *several* nodes?

File: include/Flow.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    Flow() = default;
+    ~Flow() = default;
+
+   /**
+    * @brief Send packets from every server sender in the amount of CWND size.
+    */
+    void Send();
+
+    /**
+    * @brief Last ID given to network device.
```
Comment (drgulevich) [inline include/Flow.hpp:25]: Which network device? Should this be in a Flow class?

File: src/Flow.cpp [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    for (std::uint32_t distance : distances) {
+        source_nodes.emplace_back(++last_given_id, distance);
+    }
+}
+
+void Flow::Send() {
+    for (ServerSender& s: source_nodes) {
+        for (std::uint64_t i = 0; i < s.GetCWNDSize(); ++i) {
+            packets.emplace(s.GetID(), s.GetCurrentTime(), s.GetCurrentTime() + s.GetDistance());
+            s.IncreaseCurrentTime(s.GetDistance());
```
Comment (drgulevich) [inline src/Flow.cpp:14]: Please, write the code without getters/setters for readability.

Comment (drgulevich) [reply src/Flow.cpp:14]: The issue description does not say to implement `Flow::Send` at this moment. If you are disagree than this point should be discussed in the issue or here.

**Comment by AntoxaBarin** [comment]
I agree with all comments. Is it true that `Flow` -- is a single directed edge in a network graph, where direction is `src_node` --> `dst_node`?

File: include/Flow.hpp (hunk 4 above)
Comment (AntoxaBarin) [reply include/Flow.hpp]: No, I will move it to NetworkSimulator class.

**Comment by AntoxaBarin** [comment]
image

**Comment by drgulevich** [comment]
> I agree with all comments. Is it true that `Flow` -- is a single directed edge in a network graph, where direction is `src_node` --> `dst_node`?

You can think of a *flow* as a directed graph (https://en.wikipedia.org/wiki/Flow_network). The flow contains information about source, destination and path (or several paths). A path can be a multiple-hop path, i.e. pass through multiple edges of the graph.

File: include/Flow.hpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include "Packet.hpp"
+#include "NetworkDevice.hpp"
+
+#include <cstdint>
+#include <memory>
+
+/**
+ * Packet flow between source and destination nodes (network devices).
```
Comment (drgulevich) [inline include/Flow.hpp:10]: Nit: "Network flow between the source and destination nodes"

**Comment by drgulevich** [comment]
> The PR should say "Closes Issue #9." This way it will close the issue automatically after the merge is complete.

What I meant here is that PR should start with "Closes Issue #9", followed by a PR description (a brief info about what has been done).

File: include/Flow.hpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include "Packet.hpp"
+#include "NetworkDevice.hpp"
+
+#include <cstdint>
+#include <memory>
```
Comment (drgulevich) [inline include/Flow.hpp:7]: `Packet.hpp` and `cstdint` are not used in this file.

**Comment by drgulevich** [comment]
> The flow contains information about source, destination and path (or several paths).

Update: for the sake of simplicity let us assume that *1 flow corresponds to 1 and only 1 path*. If we need to send data between the same source and destination across several paths (e.g. in the context of load balancing) than *several flows* have to be created for within the same (user-level) *session*.

**Review by drgulevich — COMMENTED** [review]
This seems too complex. I have updated the `README` in the main branch. Please, follow this guidance for the flow:

    Flow — источник пакетов. Он планирует регулярные события для генерации новых пакетов.
    Когда наступает время отправки нового пакета, Flow вызывает Sender::Enqueue(packet).

https://github.com/Cloud-Storage-Team/algnet/blob/main/README.md#flow

**Comment by drgulevich** [comment]
@AntoxaBarin, please, complete this issue when ready following the guidance above. Thanks

File: include/Flow.hpp [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+public:
+    explicit Flow(const std::vector<std::shared_ptr<NetworkDevice>>& path, const std::vector<std::uint32_t>& distances_ns);
+
+    void Send();
+
+    std::uint32_t id;
+    std::vector<std::shared_ptr<NetworkDevice>> path{};
+    std::vector<std::uint32_t> distances_ns{};
+
+    double interval_ns;
```
Comment (drgulevich) [inline include/Flow.hpp:19]: What is `interval_ns` for?

Comment (AntoxaBarin) [reply include/Flow.hpp]: I renamed it to `packet_generation_interval_ns`. It is a event delay time in `Flow::Send()`. 

A new packet is enqueued every `packet_generation_interval_ns` nanoseconds.

File: include/Flow.hpp [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+#include <queue>
+
+class Flow {
+public:
+    explicit Flow(const std::vector<std::shared_ptr<NetworkDevice>>& path, const std::vector<std::uint32_t>& distances_ns);
+
+    void Send();
+
+    std::uint32_t id;
+    std::vector<std::shared_ptr<NetworkDevice>> path{};
```
[the remark this answers is not in the document]
Comment (AntoxaBarin) [reply include/Flow.hpp:17]: Currently, the `path` is only being used by NetworkSimulator to build routing table. We can replace `path` with source and destination and add method `AddRoute` (or `AddLink`) in NetworkSimulator to configure network path.

File: include/Flow.hpp (hunk 8 above)
Comment (drgulevich) [reply include/Flow.hpp]: Could you, please, draw a schematics of the network device you have in mind?

**Comment by drgulevich** [comment]
Please, convert this to PR and ping me when ready.

File: include/Flow.hpp (hunk 8 above)
Comment (AntoxaBarin) [reply include/Flow.hpp]: image

Could you please explain why the delta is needed?

File: src/Flow.cpp [hunk 10]
```diff
@@ -0,0 +1,19 @@
+#include "Flow.hpp"
+#include "NetworkSimulator.hpp"
+
+#include <stdexcept>
+
+Flow::Flow(const std::vector<std::shared_ptr<NetworkDevice>>& path,
```
Comment (drgulevich) [inline src/Flow.cpp:6]: Sorry that I confused you here with the flow paths. As soon as we now have a routing table in the PR https://github.com/Cloud-Storage-Team/algnet/pull/33, I think, there is no need to set path in the constructor anymore, but only *source* and *destination*. Basically, the flow should be a direct analog of the *network socket pair*. Creating/deleting flows is equivalent to opening/closing a socket.

File: src/Flow.cpp [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+        packet_generation_interval_ns(packet_generation_interval_ns) {
+    if (path.size() < 2) {
+        throw std::runtime_error("Error: Incorrect flow path.");
+    }
+    if (distances_ns.size() != path.size() - 1) {
+        throw std::runtime_error("Error: Incorrect flow distances.");
+    }
+    this->path = path;
+    this->distances_ns = distances_ns;
+    id = Flow::last_given_flow_id++;
```
Comment (drgulevich) [inline src/Flow.cpp:18]: Thinking of a flow as a *socket pair*, I suppose the flow id can be hashed from the ids of the source and destination?

File: src/Flow.cpp [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+           std::uint64_t packet_generation_interval_ns):
+        packet_generation_interval_ns(packet_generation_interval_ns) {
+    if (path.size() < 2) {
+        throw std::runtime_error("Error: Incorrect flow path.");
+    }
+    if (distances_ns.size() != path.size() - 1) {
+        throw std::runtime_error("Error: Incorrect flow distances.");
+    }
+    this->path = path;
+    this->distances_ns = distances_ns;
```
Comment (drgulevich) [inline src/Flow.cpp:17]: I doubt that a flow should know about *distances*. There will be a separate data structure for links (#39).

File: include/Flow.hpp (hunk 8 above)
Comment (drgulevich) [reply include/Flow.hpp]: Sorry, this sketch was meant to be here https://github.com/Cloud-Storage-Team/algnet/pull/33

Comment (drgulevich) [reply include/Flow.hpp]: > Could you please explain why the delta is needed?

This is the exact formula which account that `t_out-t_in` may not be multiple of `packet_processing_delay`. Assuming that `packet_processing_delay` is small compared to other delays, `delta` can be neglected.

File: include/Flow.hpp [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+                  const std::vector<std::uint32_t>& distances_ns,
+                  std::uint64_t packet_generation_interval_ns);
+
+    void Send();
+
+    std::uint32_t id;
+    std::vector<std::shared_ptr<NetworkDevice>> path{};
+    std::vector<std::uint32_t> distances_ns{};
+
+    std::uint64_t packet_generation_interval_ns = 0;
```
Comment (drgulevich) [inline include/Flow.hpp:21]: Nit: `m_packet_generation_interval_ns` for member variables

File: include/Flow.hpp (hunk 8 above)
Comment (drgulevich) [reply include/Flow.hpp]: Actually, the formula that use https://github.com/Cloud-Storage-Team/algnet/pull/34#discussion_r1895677184 takes into account this `delta` naturally: it sits in `predicted_completion_time - now`, which may not be multiple of the number of packets.

**Review by drgulevich — APPROVED** [review]
LGTM. Thanks for the changes. This can now be merged.

File: include/Flow.hpp [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+#include <queue>
+
+class Flow {
+public:
+    explicit Flow(std::shared_ptr<NetworkDevice> sender, std::shared_ptr<NetworkDevice> receiver,
+                  std::uint64_t packet_generation_interval_ns);
+
+    void Send();
+
+    std::shared_ptr<NetworkDevice> m_sender;
```
Comment (drgulevich) [inline include/Flow.hpp:16]: Should these members be *private*?

**Review by drgulevich — APPROVED** [review]
Approved subjected to the correction above.

File: include/Flow.hpp (hunk 14 above)
Comment (AntoxaBarin) [reply include/Flow.hpp:16]: Yes, I think every member, except constructor, should be private

Comment (AntoxaBarin) [reply include/Flow.hpp:16]: Yes, I think every member, except constructor, should be private

**Review by drgulevich — APPROVED** [review]
Thanks for the changes

**Pull request merged**